File: c++/LeetCodeSolutions/TrappingRainwaterProblem.cpp
// LEETCODE PROBLEM :  Trapping Rain Water

// LINK TO THE PROBLEM : https://leetcode.com/problems/trapping-rain-water/

// Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.


#include <iostream>
using namespace std;
#include <vector>

class Solution {
public:
    int trap(vector<int>& height) {
        int water = 0, var;
        int n = height.size();
        std::vector<int> left(n), right(n);

        left[0] = height[0];
        for (int i = 1; i < n; i++) {
            left[i] = std::max(left[i - 1], height[i]);
        }

        right[n - 1] = height[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            right[i] = std::max(right[i + 1], height[i]);
        }

        for (int i = 1; i < n - 1; i++) {
            var = std::min(left[i - 1], right[i + 1]);
            if (var > height[i]) {
                water += var - height[i];
            }
        }

        return water;
    }
};

int main() {

    vector<int> height = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    Solution solution;
    int trappedWater = solution.trap(height);
    cout << "Trapped water: " << trappedWater << std::endl;

    return 0;
}

File: JAVA/trapping_rainwater_problem.java
/*
 Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.

 Example 1:
 Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
 Output: 6
 Explanation: The above elevation map (black section) is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped.

 Example 2:

 Input: height = [4,2,0,3,2,5]
 Output: 9
 

 Constraints:

 n == height.length
 1 <= n <= 2 * 104
 0 <= height[i] <= 105
 */

 import java.util.Scanner;
 class Trapping_Rainwater{
    public static int Trap(int arr[]){
        //left max Auxilary Array
        int n=arr.length;
        int leftMax[]=new int[n];
        leftMax[0]=arr[0];
        for(int i=1;i<n;i++){
            leftMax[i]=Math.max(arr[i], leftMax[i-1]);
        }
        //right max Auxilary Array
        int rightMax[]=new int[n];
        rightMax[n-1]=arr[n-1];
        for(int i=n-2;i>=0;i--){
            rightMax[i]=Math.max(arr[i], rightMax[i+1]);
        }
        //main loop
        int sum=0;
        for(int i=0;i<n;i++){
            int water_level=Math.min(rightMax[i], leftMax[i]);
            sum+=water_level-arr[i];
        }
        return sum;
    }

    public static void main(String[] args) {
        int arr[]={4,2,0,6,3,2,5};
        System.out.println("Rainwater Trapped:- "+Trap(arr));
    }
 }